server/src/main/java/com/cloud/vm/UserVmManagerImpl.java

```diff
@@ -2751,27 +2749,55 @@ public UserVm destroyVm(DestroyVMCmd cmd) throws ResourceUnavailableException, C
         // check if VM exists
         UserVmVO vm = _vmDao.findById(vmId);
 
-        if (vm == null) {
+        if (vm == null || vm.getRemoved() != null) {
             throw new InvalidParameterValueException("unable to find a virtual machine with id " + vmId);
         }
 
+        if (vm.getState() == State.Destroyed || vm.getState() == State.Expunging) {
+            s_logger.trace("Vm id=" + vmId + " is already destroyed");
+            return vm;
+        }
+
         // check if there are active volume snapshots tasks
         s_logger.debug("Checking if there are any ongoing snapshots on the ROOT volumes associated with VM with ID " + vmId);
         if (checkStatusOfVolumeSnapshots(vmId, Volume.Type.ROOT)) {
             throw new CloudRuntimeException("There is/are unbacked up snapshot(s) on ROOT volume, vm destroy is not permitted, please try again later.");
         }
         s_logger.debug("Found no ongoing snapshots on volume of type ROOT, for the vm with id " + vmId);
 
+        List<VolumeVO> volumes = getVolumesFromIds(cmd);
+
+        checkForUnattachedVolumes(vmId, volumes);
+        validateVolumes(volumes);
+        detachVolumesFromVm(volumes);
```

Comment:
Shouldn't we detach after the VM has been stopped? As the detach might not work, let's say libvirt (KVM) hangs or something else.

If the VM is stopped the detach is a lot easier and you can remove them then.

Reply:
We don't stop the VM for a regular detach cmd? This has been tested on KVM (with a small and large amount of volumes) and it worked fine. I'm happy to make the change if required though

Reply:
@wido Can you explain the risk here? It seems no different than a regular detach. Do these fail a lot?

Reply:
@wido this change has been made. please give the PR another review